gateway/src/apicast/policy/rewrite_url_captures/named_args_matcher.lua

```diff
@@ -0,0 +1,115 @@
+local re_split = require('ngx.re').split
+local insert = table.insert
+local unpack = unpack
+local ipairs = ipairs
+local setmetatable = setmetatable
+
+local _M = {}
+
+local mt = { __index = _M }
+
+-- Returns a list of named args extracted from a match_rule.
+-- For example, for the rule /{abc}/{def}?{ghi}=1, it returns this list:
+-- { "abc", "def", "ghi" }.
+local function extract_named_args(match_rule)
+  local iterator, err = re_gmatch(match_rule, [[\{(.+?)\}]], 'oj')
+
+  if not iterator then
+    return nil, err
+  end
+
+  local named_args = {}
+
+  while true do
+    local m, err_iter = iterator()
+    if err_iter then
+      return nil, err_iter
+    end
+
+    if not m then
+      break
+    end
+
+    insert(named_args, m[1])
+  end
+
+  return named_args
+end
+
+-- Rules contain {} for named args. This function replaces those with "()" to
+-- be able to capture those args when matching the regex.
+local function transform_rule_to_regex(match_rule)
+  return re_gsub(
+    match_rule,
+    [[\{.+?\}]],
+    [[([\w-.~%!$$&'()*+,;=@:]+)]], -- Same as in the MappingRule module
+    'oj'
+  )
+end
+
+-- Transforms a string representing the args of a query like:
+-- "a=1&b=2&c=3" into a table: { a = "1", b = "2", c = "3" }.
+local function string_params_to_table(string_params)
+  local res = {}
+
+  local params_split = re_split(string_params, '&')
+
+  for _, param in ipairs(params_split) do
+    local name, val = unpack(re_split(param, '='))
+    res[name] = val
+  end
+
+  return res
+end
+
+local function replace_in_template(args, vals, template)
+  local res = template
+
+  for i, arg in ipairs(args) do
+    res = re_gsub(res, "{" .. arg .. "}", vals[i], 'oj')
+  end
+
+  return res
+end
+
+--- Initialize a NamedArgsMatcher
+-- @tparam string match_rule Rule to be matched and that contains named args
+--   with "{}". For example: "/{var_1}/something/{var_2}".
+-- @tparam string template Template in which the named args matched will be
+--   replaced. For example: "/v2/something/{var_1}?my_arg={var_2}".
+function _M.new(match_rule, template)
+  local self = setmetatable({}, mt)
+
+  self.named_args = extract_named_args(match_rule)
+  self.regex_rule = transform_rule_to_regex(match_rule)
+  self.template = template
+
+  return self
+end
+
+--- Match a path
+-- @tparam string path The path of the URL
+-- @treturn boolean True if there's a match, false otherwise
+-- @treturn string The new path. If there's a match
+-- @treturn table The new args. If there's a match
+-- Note: this method generates a new url and new query args when there is a
+-- match, but does not modify the current ones.
+function _M:match(path)
+  local matches = re_match(path, self.regex_rule, 'oj')
+
+  if not matches or #self.named_args ~= #matches then
+    return false
+  end
+
+  local replaced_template = replace_in_template(
+    self.named_args, matches, self.template)
+
+  local uri, raw_params = unpack(re_split(replaced_template, '\\?'))
```

Comment:
`unpack` can't be JITed so this should not be in the hot path.

Reply:
And instead of `'\\?'` you can use `[[\?]]`, but that might be just my preference.

Reply:
Fixed the `unpack`.
I don't have a strong preference on `\\?` vs `[[\?]]` so I changed it.